client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt

```diff
@@ -0,0 +1,203 @@
+package app.cash.backfila.client.misk
+
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.KeyRange
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import java.util.ArrayDeque
+
+//TODO(mikepaw) do I need to do anything to make this more explicitly not thread safe?
+class EmbeddedBackfillRun<BackfillType: Backfill> (
+  val backfill: BackfillType,
+  val dryRun: Boolean,
+  val parameters: Map<String, ByteString>,
+  val rangeStart: String?,
+  val rangeEnd: String?,
+
+  var batchSize: Long = 100L,
+  var scanSize: Long = 10000L,
+  var computeCountLimit: Long = 1L
+  // Purposely not adding threads, sleeping or backoff schedule, may reevaluate that later
+) {
+  val prepareBackfillResponse: PrepareBackfillResponse
+
+  private val precomputeProgress: Map<String, InstanceCursor>
+  var precomputeMatchingCount: Long = 0L
+    private set
+  var precomputeScannedCount: Long = 0L
+    private set
+
+  private val scanProgress: Map<String, InstanceCursor>
+  val instanceProgressSnapshot: Map<String, InstanceCursorSnapshot>
+    get() = scanProgress.map { Pair(it.key, InstanceCursorSnapshot(it.value)) }.toMap()
+
+  private val batchesToRun: ArrayDeque<BatchInfo> = ArrayDeque()
+  val batchesToRunSnapshot: List<BatchInfo>
+    get() = batchesToRun.toList()
+
+  /** Prepares a backfill for a run. */
+  init {
+    var response = backfill.prepareBackfill(PrepareBackfillRequest.Builder()
+        .backfill_name(backfill.javaClass.toString())
+        .range(KeyRange(rangeStart?.encodeUtf8(), rangeEnd?.encodeUtf8()))
+        .parameters(parameters)
+        .dry_run(dryRun)
+        .build())
+    prepareBackfillResponse = response
+    precomputeProgress = response.instances.map { Pair(
+        it.instance_name,
+        InstanceCursor(it.instance_name,
+            it.backfill_range)) }.toMap()
+    scanProgress = response.instances.map { Pair(
+        it.instance_name,
+        InstanceCursor(it.instance_name,
+            it.backfill_range)) }.toMap()
+  }
+
+  // Precomputing Methods
+  fun precomputeScan(): GetNextBatchRangeResponse {
+    val notDone = precomputeProgress.filter { !it.value.done }
+    check(notDone.isNotEmpty()) {"Nothing left to precompute"}
```

Comment:
nit: add spaces inside `{}`
```suggestion
    check(notDone.isNotEmpty()) { "Nothing left to precompute" }
```